rr_evgp/src/binary_bayes_filter_obstacle_layer.cpp

```diff
@@ -0,0 +1,250 @@
+                range = scan_range_;
+            }
+            double wx = lidar_x_ + range * std::cos(lidar_yaw_ + angle);
+            double wy = lidar_y_ + range * std::sin(lidar_yaw_ + angle);
+            angle += scan_->angle_increment;
+
+            double fudge = 2 * resolution;
+            if (wx - fudge < *min_x)
+                *min_x = std::max(costmap->getOriginX(), wx - fudge);
+            if (wx + fudge > *max_x)
+                *max_x = std::min(costmap->getOriginX() + costmap->getSizeInMetersX(), wx + fudge);
+            if (wy - fudge < *min_y)
+                *min_y = std::max(costmap->getOriginY(), wy - fudge);
+            if (wy + fudge > *max_y)
+                *max_y = std::min(costmap->getOriginY() + costmap->getSizeInMetersY(), wy + fudge);
+        }
+    }
+
+    void updateCosts(costmap_2d::Costmap2D& master_grid, int min_cell_x, int min_cell_y, int max_cell_x,
+                     int max_cell_y) override {
+        if (!enabled_ || !scan_) {
+            return;
+        }
+
+        const double resolution = master_grid.getResolution();
+        const double world_min_x = master_grid.getOriginX();
+        const double world_max_x = world_min_x + master_grid.getSizeInMetersX();
+        const double world_min_y = master_grid.getOriginY();
+        const double world_max_y = world_min_y + master_grid.getSizeInMetersY();
+
+        for (int mx = min_cell_x; mx < max_cell_x; mx++) {
+            for (int my = min_cell_y; my < max_cell_y; my++) {
+                updates_[master_grid.getIndex(mx, my)] = -1;
+            }
+        }
+
+        double angle = scan_->angle_min;
+        for (double raw_range : scan_->ranges) {
+            bool is_hit = (!std::isinf(raw_range) && raw_range <= scan_range_);
+            double range = is_hit ? raw_range : scan_range_;
+            double trace_range = range + (resolution * 0.7);
+            double wx = lidar_x_ + trace_range * std::cos(lidar_yaw_ + angle);
+            double wy = lidar_y_ + trace_range * std::sin(lidar_yaw_ + angle);
+
+            if (wx < world_min_x || wx >= world_max_x || wy < world_min_y || wy >= world_max_y) {
+                ROS_WARN_THROTTLE(1.0, "trying to scan outside the map");
+                continue;
+            }
+
+            const double step_size = resolution * 0.25;
+            const double step_x = step_size * (wx - lidar_x_) / range;
+            const double step_y = step_size * (wy - lidar_y_) / range;
+            int mx, my;
+            double d = 0;
+            double march_x = lidar_x_;
+            double march_y = lidar_y_;
+            const double last_clear = range - resolution;
+            while (d < last_clear) {
+                master_grid.worldToMapNoBounds(march_x, march_y, mx, my);
+                updates_[master_grid.getIndex(mx, my)] = 0;
+                d += step_size;
+                march_x += step_x;
+                march_y += step_y;
+            }
```

Comment:
Interesting that you do it this way. How is the performance like?

Reply:
I haven't profiled this method vs costmap_2d's Bresenham implementation, but overall this layer has the same or less CPU usage than ObstacleLayer. My raytracing loop executes more times but has no branches